### ISSUE TITLE
Figuring out 'txInfoSignatories' format

### ISSUE THREAD

**Issue by hasufell** [input-output-hk/plutus-apps#16]
I have a validator that is fairly straight forward:

```hs
validateSpend :: ValidatorType Burner
validateSpend (MyDatum addrHash) _myRedeemerValue ScriptContext { scriptContextTxInfo = TxInfo { txInfoSignatories = [addr] } } =
   addrHash == sha3_256 (getPubKeyHash addr)
validateSpend _ _  _ = traceError "Expecting exactly one signatory."
```

I compiled this plutus script, built a transaction manually and can lock funds successfully. I do this via wallet api, cardano-cli and other cli tools, because plutus isn't complete yet.

Now I want to lock funds for myself, so I can redeem them, but in order to do that I need to understand the **exact** format/encoding/hashing of `txInfoSignatories`. I've been searching for hours and tried many variations, using the wallet pubkey, account pubkey, address pubkey, hashing with Blake2b_224 etc. pp.

I find it hard to find in the documentation 1) what exact pubkey is this and 2) what exact hash/format is this and how can I create this hash from the command line given that I have full access to the wallet that signed the transaction.

All my attempts so far redeeming the funds ended with:

```
Command failed: transaction build  Error: The following scripts have execution failures:
the script for transaction input 0 (in the order of the TxIds) failed with The Plutus script evaluation failed: An error has occurred:  User error:
The provided Plutus code called 'error'.
```

Thanks.

**Comment by hasufell** [comment]
@luigy still doesn't work for me. With latest cardano-node etc I get:


```
Command failed: transaction build  Error: The following scripts have execution failures:
the script for transaction input 0 (in the order of the TxIds) failed with:
The Plutus script evaluation failed: An error has occurred:  User error:
The provided Plutus code called 'error'.
Caused by: [ (force (builtin headList)) (con list (data) []) ]
```

The full code is here: https://git.io/JPD3v

**Comment by hasufell** [comment]
Ok, I was able to track down the issue further.

As described above, we use a redeemer dummy type, because whether redeeming works or not is supposed to be discovered by the `txInfoSignatories`, not the redeemer datum:

```hs
-- | Redeemer holds no data, since the address is taken from the context.
newtype MyRedeemer = MyRedeemer { _nothing :: () }
PlutusTx.makeLift ''MyRedeemer
PlutusTx.unstableMakeIsData ''MyRedeemer
```

I followed down the code of cardano-api in the node, which apparently does the json encoding/decoding: https://github.com/input-output-hk/cardano-node/blob/d95ef21062a50acf8f35f1f5d94f8a8a70f7510b/cardano-api/src/Cardano/Api/ScriptData.hs#L431

In a repl, I was then able to generate the the json from the node code directly:

```
> Data.Aeson.encode $ scriptDataToJson ScriptDataJsonDetailedSchema $ (fromPlutusData $ builtinDataToData $ toBuiltinData (MyRedeemer ()))
"{\"constructor\":0,\"fields\":[{\"constructor\":0,\"fields\":[]}]}"
```

In my code, I tried the json values `""`, `"{\"constructor\":0,\"fields\":[]}"` and some other combinations, but none of them match the above. After fixing that, I could submit the redeem transaction and got a different error:

```
the script for transaction input 0 (in the order of the TxIds) failed with:
The Plutus script evaluation failed: An error has occurred:  User error:
The provided Plutus code called 'error'.
Script debugging logs: owner has not signed
PT5
```

This is much better. Now the script actually tells me that it couldn't find the signature in the `txInfoSignatories`. After further tests, these apparently evaluate to an empty list, although I'm passing `--required-signer` to `cardano-cli transaction build`.

So that's at least some progress.

**Comment by luigy** [comment]
@hasufell nice, well done. With the new error you might be bumping into a bug that was fixed by https://github.com/input-output-hk/cardano-node/pull/3319 in which `PaymentExtendedKeys` were being dropped from requiredSignatories list. I skimmed through how your keys are being converted and it might be the case
https://github.com/hasufell/proof-of-burn-cardano/blob/26bfa911a371d37e69812321bc787df933e03d92/docker/cardano-cli/wallet.sh#L343
https://github.com/input-output-hk/cardano-node/blob/d95ef21062a50acf8f35f1f5d94f8a8a70f7510b/cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs#L533

**Comment by AngelCastilloB** [comment]
> Ok, I was able to track down the issue further.
> 
> As described above, we use a redeemer dummy type, because whether redeeming works or not is supposed to be discovered by the `txInfoSignatories`, not the redeemer datum:
> 
> ```haskell
> -- | Redeemer holds no data, since the address is taken from the context.
> newtype MyRedeemer = MyRedeemer { _nothing :: () }
> PlutusTx.makeLift ''MyRedeemer
> PlutusTx.unstableMakeIsData ''MyRedeemer
> ```
> 
> I followed down the code of cardano-api in the node, which apparently does the json encoding/decoding: https://github.com/input-output-hk/cardano-node/blob/d95ef21062a50acf8f35f1f5d94f8a8a70f7510b/cardano-api/src/Cardano/Api/ScriptData.hs#L431
> 
> In a repl, I was then able to generate the the json from the node code directly:
> 
> ```
> > Data.Aeson.encode $ scriptDataToJson ScriptDataJsonDetailedSchema $ (fromPlutusData $ builtinDataToData $ toBuiltinData (MyRedeemer ()))
> "{\"constructor\":0,\"fields\":[{\"constructor\":0,\"fields\":[]}]}"
> ```
> 
> In my code, I tried the json values `""`, `"{\"constructor\":0,\"fields\":[]}"` and some other combinations, but none of them match the above. After fixing that, I could submit the redeem transaction and got a different error:
> 
> ```
> the script for transaction input 0 (in the order of the TxIds) failed with:
> The Plutus script evaluation failed: An error has occurred:  User error:
> The provided Plutus code called 'error'.
> Script debugging logs: owner has not signed
> PT5
> ```
> 
> This is much better. Now the script actually tells me that it couldn't find the signature in the `txInfoSignatories`. After further tests, these apparently evaluate to an empty list, although I'm [passing `--required-signer` to `cardano-cli transaction build`](https://github.com/hasufell/proof-of-burn-cardano/blob/26bfa911a371d37e69812321bc787df933e03d92/docker/cardano-cli/wallet.sh#L489).
> 
> So that's at least some progress.

I was also stuck with the same issue (wrong redeemer format), we are also not using the redeemer and were passing [] as a dummy redeemer value. For the longest time, we were looking into the Datums because we were not using the redeemer so we didn't think the problem was there, if not were for your finding we would still be looking. 

This was the error we were getting:

```
The Plutus script evaluation failed: An error has occurred:  User error:
The provided Plutus code called 'error'.
Caused by: [ (builtin unConstrData) (con data #80) ]
```

**Comment by hasufell** [comment]
> @hasufell nice, well done. With the new error you might be bumping into a bug that was fixed by [input-output-hk/cardano-node#3319](https://github.com/input-output-hk/cardano-node/pull/3319) in which `PaymentExtendedKeys` were being dropped from requiredSignatories list. I skimmed through how your keys are being converted and it might be the case https://github.com/hasufell/proof-of-burn-cardano/blob/26bfa911a371d37e69812321bc787df933e03d92/docker/cardano-cli/wallet.sh#L343 https://github.com/input-output-hk/cardano-node/blob/d95ef21062a50acf8f35f1f5d94f8a8a70f7510b/cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs#L533

I rebuilt with that commit and it indeed fixed it and I could submit the redeem transaction to testnet and it apparently validated correctly and I got my funds back.

So, to reiterate (I'll probably write an article on it too):

1. feed your data (either Datum or Redeemer) to the function `Data.Aeson.encode . scriptDataToJson ScriptDataJsonDetailedSchema . fromPlutusData . builtinDataToData . toBuiltinData` to figure out the final encoding
2. make sure to use `cardano-cli transaction build --tx-in-redeemer-file $FILE1 --tx-in-datum-file $FILE2` with files that contain the json content from the step above when constructing transactions (details: there are two json schemas for decoding... the extended one `ScriptDataJsonDetailedSchema` is used when reading files as opposed to direct json strings and has better errors in my experience)
3. make sure to run at least with cardano-cli commit `d95ef21062a50acf8f35f1f5d94f8a8a70f7510b`, so `txInfoSignatories` are not dropped
4. if the script fails without a `PT<int>` error, it didn't even manage to evaluate, but failed during decoding
5. generating wallet pubkey (e.g. for use in datum or redeemer) can be done like so: `cardano-cli address key-hash --payment-verification-key "$(echo "<your recovery phrase>" | cardano-address key from-recovery-phrase Shelley | cardano-address key child "1852H/1815H/0H/0/0" | cardano-address key public --with-chain-code)"` (make sure to include it via `cardano-cli transaction build --required-signer=key.skey`, which requires the signing key of the same derivation path... I also had to set `--witness-override 2`, not sure if that's still needed)

Thanks for all your help.

----

@AngelCastilloB glad to hear
